pkg/applier/applier.go

```diff
@@ -589,83 +602,52 @@ func (a *supervisor) applyInner(ctx context.Context, objs []client.Object) (map[
 			} else {
 				klog.V(1).Info(e.WaitEvent)
 			}
-			a.addError(eh.processWaitEvent(e.WaitEvent, s.WaitEvent, objStatusMap))
+			if err := eh.processWaitEvent(e.WaitEvent, syncStats.WaitEvent, objStatusMap); err != nil {
+				a.sendErrorEvent(err, eventHandler)
+			}
 		case event.ApplyType:
 			if e.ApplyEvent.Error != nil {
 				klog.Info(e.ApplyEvent)
 			} else {
 				klog.V(1).Info(e.ApplyEvent)
 			}
-			a.addError(eh.processApplyEvent(ctx, e.ApplyEvent, s.ApplyEvent, objStatusMap, unknownTypeResources, resourceMap))
+			if err := eh.processApplyEvent(ctx, e.ApplyEvent, syncStats.ApplyEvent, objStatusMap, unknownTypeResources, resourceMap); err != nil {
+				a.sendErrorEvent(err, eventHandler)
+			}
 		case event.PruneType:
 			if e.PruneEvent.Error != nil {
 				klog.Info(e.PruneEvent)
 			} else {
 				klog.V(1).Info(e.PruneEvent)
 			}
-			a.addError(eh.processPruneEvent(ctx, e.PruneEvent, s.PruneEvent, objStatusMap))
+			if err := eh.processPruneEvent(ctx, e.PruneEvent, syncStats.PruneEvent, objStatusMap); err != nil {
+				a.sendErrorEvent(err, eventHandler)
+			}
 		default:
 			klog.Infof("Unhandled event (%s): %v", e.Type, e)
 		}
 	}
 
-	gvks := make(map[schema.GroupVersionKind]struct{})
-	for _, resource := range objs {
-		id := core.IDOf(resource)
-		if _, found := unknownTypeResources[id]; found {
-			continue
-		}
-		gvks[resource.GetObjectKind().GroupVersionKind()] = struct{}{}
-	}
-
-	errs := a.Errors()
-	if errs == nil {
-		klog.V(4).Infof("Apply completed without error: all resources are up to date.")
-	}
-	if s.Empty() {
-		klog.V(4).Infof("Applier made no new progress")
-	} else {
-		klog.Infof("Applier made new progress: %s", s.String())
-		objStatusMap.Log(klog.V(0))
-	}
-	return gvks, errs
+	return objStatusMap, syncStats
 }
 
-// Errors returns the errors encountered during the last apply or current apply
-// if still running.
-// Errors implements the Applier and Destroyer interfaces.
-func (a *supervisor) Errors() status.MultiError {
-	a.errorMux.RLock()
-	defer a.errorMux.RUnlock()
-
-	// Return a copy to avoid persisting caller modifications
-	return status.Append(nil, a.errs)
+func (a *supervisor) sendErrorEvent(err error, eventHandler func(Event)) {
```

Comment:
The `sendErrorEvent` function doesn't need to be a member of the supervisor.

Reply:
fixed.